src/Microsoft.Extensions.ObjectPool/DefaultObjectPool.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.Extensions.ObjectPool
+{
+    public class DefaultObjectPool<T> : ObjectPool<T> where T : class, new()
+    {
+        private readonly T[] _items;
+
+        public DefaultObjectPool()
+            : this(Environment.ProcessorCount * 2)
+        {
+        }
+
+        public DefaultObjectPool(int maximumRetained)
+        {
+            _items = new T[maximumRetained];
+        }
+
+        public override T Get()
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i];
+                if (item != null && Interlocked.CompareExchange(ref _items[i], null, item) == item)
+                {
+                    return item;
+                }
+            }
+
+            return new T();
+        }
+
+        public override void Return(T obj)
+        {
+            OnObjectReturned(obj);
+
+            for (var i = 0; i < _items.Length; i++)
```

Comment:
At the risk of over-engineering - would it be worth storing the last successful "return" index, and have this loop start there and wrap around rather than always starting from 0? Seems like that would tend to pass over the "least recently returned" slots first, which would tend to hit a null with fewer iterations.

Same thought probably applies to Get, though you'd probably want two different "last successful" indexes, one for Get and another for Return...